src/main/java/com/bushidowallet/core/bitcoin/tx/Transaction.java

```diff
@@ -1,356 +1,395 @@
+package com.bushidowallet.core.bitcoin.tx;
+
+import com.bushidowallet.core.bitcoin.Address;
+import com.bushidowallet.core.bitcoin.bip32.ECKey;
+import com.bushidowallet.core.bitcoin.bip32.Hash;
+import com.bushidowallet.core.bitcoin.script.Script;
+import com.bushidowallet.core.bitcoin.tx.input.Input;
+import com.bushidowallet.core.bitcoin.tx.input.MultiSigScriptHashInput;
+import com.bushidowallet.core.bitcoin.tx.input.PublicKeyHashInput;
+import com.bushidowallet.core.bitcoin.tx.output.Output;
+import com.bushidowallet.core.bitcoin.util.ByteReader;
+import com.bushidowallet.core.bitcoin.util.ByteWriter;
+import com.bushidowallet.core.crypto.util.ByteUtil;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by Jesion on 2015-03-31.
+ */
+public class Transaction {
+
+    private static int CURRENT_VERSION = 1;
+    private static int DEFAULT_NLOCKTIME = 0;
+    private static int MAXIMUM_EXTRA_SIZE = 4 + 9 + 9 + 4;
+    private static int CHANGE_OUTPUT_MAX_SIZE = 20 + 4 + 34 + 4;
+    private static long FEE_PER_KB = 10000;
+
+    public int version;
+
+    public List<Input> inputs;
+
+    public List<Output> outputs;
+
+    public int nLockTime;
+
+    private Script changeScript;
+
+    private int changeIndex;
+
+    private long fee;
+
+    public Transaction() {
+
+        inputs = new ArrayList<Input>();
+        outputs = new ArrayList<Output>();
+
+        version = CURRENT_VERSION;
+        nLockTime = DEFAULT_NLOCKTIME;
+
+        changeIndex = -1;
+        fee = -1;
+    }
+
+    public void sign(List<ECKey> keys) throws Exception {
+        sign(keys, TransactionSignature.SIGHASH_ALL);
+    }
+
+    /**
+     * Signs this Tx with a collection of private keys provided
+     *
+     * @param keys - collection of private keys used for signing
+     * @param sigType - see Signature
+     * @throws Exception
+     */
+    public void sign(List<ECKey> keys, int sigType) throws Exception {
+        for (int i = 0; i < keys.size(); i++) {
+            sign(keys.get(i), sigType);
+        }
+    }
+
+	public void sign(ECKey key, int sigType) throws Exception {
+		if (hasAllUtxoInfo()) {
+			List<TransactionSignature> signatures = getSignatures(key, sigType);
+			for (int j = 0; j < signatures.size(); j++) {
+				applySignature(signatures.get(j));
+			}
+		} else {
+			throw new Exception("Inputs not complete");
+		}
+	}
+
+	/**
+	 * add by Liwl
+	 * muilt sign by one person.
+	 * 
+	 * @param key
+	 * @param sigType
+	 * @return
+	 * @throws Exception
+	 */
+	public List<TransactionSignature> signByOne(ECKey key, int sigType) throws Exception {
+		if (hasAllUtxoInfo()) {
+			return getSignatures(key, sigType);
+		} else {
+			throw new Exception("Inputs not complete");
+		}
+	}
+
+	/**
+	 * add by Liwl
+	 * appley signature by one person.
+	 * 
+	 * @param signatures
+	 * @throws Exception
+	 */
+	public void applySignatureByOne(List<TransactionSignature> signatures) throws Exception {

```

Comment:
I write a test case.
please try it.
It will occur an Exception:`java.lang.IndexOutOfBoundsException: Index: 1, Size: 0`
test code in:https://github.com/liguanliang2014/bushido-java-core/blob/master/src/test/java/com/bushidowallet/core/bitcoin/tx/TransactionTest.java
the testMutilSigTx() function.

copy code here

```
    // generate an ECDsA key
    private static ECKey generateKey() {
        try {
            //LogUtil.i("begin generate ecc key...");
            KeyPairGenerator keyGen = KeyPairGenerator.getInstance("ECDsA", new org.bouncycastle.jce.provider.BouncyCastleProvider());
            ECGenParameterSpec ecSpec = new ECGenParameterSpec("secp256k1");
            keyGen.initialize(ecSpec, new SecureRandom());
            KeyPair generateKeyPair = keyGen.generateKeyPair();
            
            BCECPrivateKey private1 = (BCECPrivateKey) generateKeyPair.getPrivate();
            BCECPublicKey public1 = (BCECPublicKey) generateKeyPair.getPublic();
            String S = private1.getS().toString(16);
            String X = public1.engineGetQ().getAffineXCoord().toBigInteger().toString(16);
            String Y = public1.engineGetQ().getAffineYCoord().toBigInteger().toString(16);
            // format string to 64 length with zero in head
            S = formatStringAdd0(S, 64);
            X = formatStringAdd0(X, 64);
            Y = formatStringAdd0(Y, 64);
            
            // public key string
            String publicKeyStr = "04" + X + Y;
            
            // set public key begin with 04
            ECKey ecKey = new ECKey(private1.getS().toByteArray(), false, true);
            return ecKey;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
    
    // test for multiple sign transaction
    // 2 of 3,people A,B,C
    // address 38bPsA6ZXfRuxFD7efVXTkQd69422uzD4B -> 1EKXAAiJDntbEK3WSVW3vTcwYJ5ciJ5kSL
    // address info ref: https://blockchain.info/address/38bPsA6ZXfRuxFD7efVXTkQd69422uzD4B
    // public key list and private key list shows below
    // is api support multiple sign tx for B&C?
    @Test
    public void testMutilSigTx() throws Exception {
    	ECKey keyOfA = generateKey();
    	ECKey keyOfB = generateKey();
    	ECKey keyOfC = generateKey();
    	
    	System.out.println("public key of A: " + keyOfA.getPublicHex());
    	System.out.println("public key of B: " + keyOfB.getPublicHex());
    	System.out.println("public key of C: " + keyOfC.getPublicHex());
    	System.out.println("private key of A: " + ByteUtil.toHex(keyOfA.getPrivate()));
    	System.out.println("private key of B: " + ByteUtil.toHex(keyOfB.getPrivate()));
    	System.out.println("private key of C: " + ByteUtil.toHex(keyOfC.getPrivate()));
    	
    	List<ECKey> publicKeyList = new ArrayList<ECKey>(3);
    	// public key of A
    	publicKeyList.add(keyOfA);
    	// public key of B
    	publicKeyList.add(keyOfB);
    	// public key of C
    	publicKeyList.add(keyOfC);
    	
    	Transaction tx = new Transaction();
    	UTXODescriptor utxoDescriptor = new UTXODescriptor();
    	utxoDescriptor.script = Script.buildScriptHashOut(Script.buildMultisigOut(publicKeyList, 2)).toString();
    	utxoDescriptor.outputIndex = 0;
        utxoDescriptor.txId = "ac89b280cc21854b82b4cc111a0e6c0d10315117b6001e3f4f3af3d2f7b2fd53";
        utxoDescriptor.satoshis = (long) 2304;
        
        List<UTXODescriptor> utxoDescriptors = new ArrayList<UTXODescriptor>(1);
        utxoDescriptors.add(utxoDescriptor);
        tx.from(utxoDescriptors, publicKeyList, 2);
        
        tx.to("1EKXAAiJDntbEK3WSVW3vTcwYJ5ciJ5kSL", 1000);
        
        // now we begin sign tx
        List<ECKey> privateKeyList = new ArrayList<ECKey>(2);
        // add private key of B and C
        privateKeyList.add(keyOfB);
        privateKeyList.add(keyOfC);
        // it will occur an Exception
        // java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
        // bug fix code in com.bushidowallet.core.bitcoin.tx.input.MultiSigScriptHashInput.addSignatureByOne
        tx.sign(privateKeyList);
    }
}
```

I don't know better way to show you the code,please test it in your project.

it can be fixed by method [addSignatureByOne](https://github.com/bushidowallet/bushido-java-core/pull/1/files/51dbda75a01ec575becec5f4e110cbf6e83cb1d4#diff-46f7541a88711a7458670b47c3548c2f).